### PR TITLE
Adding support for distributed TensorFlow.

### PR DESCRIPTION
**Pull request by vicaire**
Adding support for distributed TensorFlow.

Tests:

Local execution, non distributed:

gcloud --verbosity=debug beta ml local train --package-path=youtube-8m-private --module-name=youtube-8m-private.train -- --train_data_pattern='gs://youtube8m-ml/1/video_level/train/train*.tfrecord' --train_dir=/tmp/yt8m_train --start_new_model

Local execution, distributed:

gcloud beta ml local train --package-path=youtube-8m-private --module-name=youtube-8m-private.train --distributed --parameter-server-count=1 --worker-count=4 -- --train_data_pattern='gs://youtube8m-ml/1/video_level/train/train*.tfrecord' --train_dir=/tmp/yt8m_train --start_new_model

Running on your own machine, python 2.7:

python youtube-8m-private/train.py --train_data_pattern='/tmp/features/train*.tfrecord' --train_dir=/tmp/features/video_level_logistic_model --start_new_model

Running on your own machine, python 3:

python3 youtube-8m-vicaire/train.py --train_data_pattern='/tmp/features/train*.tfrecord' --train_dir=/tmp/features/video_level_logistic_model

Distributed execution on cloud:

BUCKET_NAME=...; JOB_NAME=yt8m_train_$(date +%Y%m%d_%H%M%S); gcloud --verbosity=debug beta ml jobs submit training $JOB_NAME --package-path=youtube-8m-private --module-name=youtube-8m-private.train --staging-bucket=$BUCKET_NAME --region=us-central1 --config=youtube-8m-private/cloudml-gpu-distributed.yaml -- --train_data_pattern='gs://youtube8m-ml/1/video_level/train/train*.tfrecord' --train_dir=$BUCKET_NAME/yt8m_train_video_level_logistic_model --start_new_model

Non-distributed execution on cloud:

BUCKET_NAME=...; JOB_NAME=yt8m_train_$(date +%Y%m%d_%H%M%S); gcloud --verbosity=debug beta ml jobs submit training $JOB_NAME --package-path=youtube-8m-private --module-name=youtube-8m-private.train --staging-bucket=$BUCKET_NAME --region=us-central1 --config=youtube-8m-private/cloudml-gpu.yaml -- --train_data_pattern='gs://youtube8m-ml/1/video_level/train/train*.tfrecord' --train_dir=$BUCKET_NAME/yt8m_train_video_level_logistic_model --start_new_model

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -143,12 +142,13 @@ def get_input_data_tensors(reader,
...
     if not files:
       raise IOError("Unable to find training files. data_pattern='" +
-                    data_pattern + "'")
-    logging.info("number of training files: " + str(len(files)))
-    filename_queue = tf.train.string_input_producer(files,
-                                                    num_epochs=num_epochs)
+                    data_pattern + "'.")
+    logging.info("Number of training files: %s.", str(len(files)))
+    filename_queue = tf.train.string_input_producer(
+        files, num_epochs=num_epochs)
```
Comment (nikhilk) [inline train.py]: Add shuffle here...

File: train.py [hunk 2]
```diff
@@ -261,148 +264,299 @@ def build_graph(reader,
...
+
+  def start_server_if_distributed(self):
+    """Starts a server if the execution is distributed."""
+
+    if self.cluster:
+      logging.info("%s: Starting trainer within cluster %s.",
+                   task_as_string(self.task), self.cluster.as_dict())
+      server = start_server(self.cluster, self.task)
+      target = server.target
+      device_fn = tf.train.replica_device_setter(
```
Comment (nikhilk) [inline train.py]: You need to tell what the ps_device is.
I don't know what merge_devices does, but I haven't used it.

File: train.py (hunk 1 above)
Comment (vicaire) [inline train.py]: Added the parameter "shuflle=True" to string_input_producer. 

It is the default but it seems that being explicit can improve readability in this case.

File: train.py (hunk 2 above)
Comment (vicaire) [inline train.py]: Added the parameter '"ps_device="/job:ps"'. It is the default value but it seems that being explicit can improve readability. 

Removing the parameter "merge_devices=True". It is also the default but this parameter is on the path to deprecation and specifying "merge_devices=False" triggers a warning.

File: train.py [hunk 3]
```diff
@@ -77,13 +80,10 @@
...
                        "How many threads to use for reading input files.")
-  flags.DEFINE_string("master", "", "TensorFlow master to use.")
-  flags.DEFINE_integer("task", 0, "Task id of the replica running the training."
-                       " 0 implies chief Supervisor.""")
-  flags.DEFINE_integer("ps_tasks", 0, """Number of tasks in the ps job.
-                       If 0 no ps job is used.""")
   flags.DEFINE_string("optimizer", "AdamOptimizer",
                       "What optimizer class to use.")
+  flags.DEFINE_bool("log_device_placement", False,
+                    "Whether device placement should be logged.")
```
Comment (LeegleechN) [inline train.py]: How about "Whether to write the device every op will run on into the logs on startup".

File: train.py [hunk 4]
```diff
@@ -143,12 +142,13 @@ def get_input_data_tensors(reader,
...
-    logging.info("number of training files: " + str(len(files)))
-    filename_queue = tf.train.string_input_producer(files,
-                                                    num_epochs=num_epochs)
+                    data_pattern + "'.")
+    logging.info("Number of training files: %s.", str(len(files)))
+    filename_queue = tf.train.string_input_producer(
+        files, num_epochs=num_epochs, shuffle=True)
     training_data = [
-        reader.prepare_reader(filename_queue) for _ in xrange(num_readers)]
+        reader.prepare_reader(filename_queue) for _ in xrange(num_readers)
```
Comment (LeegleechN) [inline train.py]: Please sync to head and test this in Python3. We are trying to maintain compatibility with both python versions now.

File: train.py (hunk 3 above)
Comment (vicaire) [inline train.py]: Changed to "Whether to write the device on which every op will run into the logs on startup."

File: train.py (hunk 4 above)
Comment (vicaire) [inline train.py]: Synced to head. 

Tested with Python3:

python3 youtube-8m-vicaire/train.py --train_data_pattern='/tmp/features/train*.tfrecord' --train_dir=/tmp/features/video_level_logistic_model
